jax/_src/tree.py

```diff
@@ -284,3 +286,45 @@ def unflatten(treedef: tree_util.PyTreeDef,
     - :func:`jax.tree.structure`
   """
   return tree_util.tree_unflatten(treedef, leaves)
+
+
+def field(*, static: bool = False, metadata: dict[str, Any] | None = None, **kwds) -> dataclasses.Field:
+  """Used to mark a jax.dataclass field as static.
+
+  See {func}`jax.tree.dataclass` for details.
+  """
+  return dataclasses.field(metadata=(metadata or {}) | {'static': static}, **kwds)
+
+
+@dataclass_transform(field_specifiers=(field, dataclasses.Field), frozen_default=True)
+def dataclass(cls: T, /, **kwargs) -> T:
```

Comment:
With this interface, it is not possible to use `@jax.tree.dataclass(...)` as a decorator (while `@dataclasses.dataclass(...)` is valid.

```suggestion
def dataclass(cls: T = None, /, **kwargs) -> T:

    if cls is None:
        return partial(dataclass, **kwargs)
```

Reply:
We will need a separate overload for the `cls: None` case returning a `Callable[..., T]`.

Reply:
Yeah I deliberately left this out in the initial draft to keep things simple.